core/trino-main/src/test/java/io/trino/execution/scheduler/TestScaledWriterScheduler.java

```diff
@@ -0,0 +1,182 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.trino.execution.scheduler;
+
+import com.google.common.collect.ImmutableList;
+import com.google.common.collect.ImmutableSet;
+import io.airlift.stats.TDigest;
+import io.airlift.units.DataSize;
+import io.airlift.units.Duration;
+import io.trino.client.NodeVersion;
+import io.trino.execution.NodeTaskMap;
+import io.trino.execution.TaskId;
+import io.trino.execution.TaskInfo;
+import io.trino.execution.TaskState;
+import io.trino.execution.TaskStatus;
+import io.trino.execution.buffer.BufferState;
+import io.trino.execution.buffer.OutputBufferInfo;
+import io.trino.execution.buffer.OutputBufferStatus;
+import io.trino.execution.scheduler.policy.TestPhasedExecutionSchedule;
+import io.trino.metadata.InMemoryNodeManager;
+import io.trino.metadata.InternalNode;
+import io.trino.operator.TaskStats;
+import io.trino.plugin.base.metrics.TDigestHistogram;
+import io.trino.util.FinalizerService;
+import org.joda.time.DateTime;
+import org.testng.annotations.Test;
+
+import java.net.URI;
+import java.util.Optional;
+import java.util.OptionalLong;
+
+import static io.airlift.concurrent.Threads.threadsNamed;
+import static io.trino.execution.scheduler.TestSourcePartitionedScheduler.createFragment;
+import static io.trino.testing.TestingSession.testSessionBuilder;
+import static java.util.concurrent.Executors.newScheduledThreadPool;
+import static java.util.concurrent.TimeUnit.MILLISECONDS;
+import static org.testng.Assert.assertEquals;
+
+public class TestScaledWriterScheduler
```

Comment:
nit: The introduction of this test could also be a separate preparatory commit. This will make the main commit more concise.

Reply:
sure.